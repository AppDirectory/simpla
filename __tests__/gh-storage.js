import GhStorage from '../src/gh-storage.js';

jest.unmock('../src/gh-storage.js');

expect.addSnapshotSerializer({
  test(val) {
    return val && Array.isArray(val) && 'headers' in val[1] && 'body' in val[1];
  },
  print(val, serialize) {
    const [ url, { method, body, headers } ] = val;
    return serialize({
      url,
      method,
      body: body && JSON.parse(body),
      auth: headers.Authorization
    });
  }
});

const MOCK_TOKEN = 'some-token';
const BLANK_RESPONSE = [ JSON.stringify(''), { status: 200 } ];
const RESPONSES = {
  getContent: [ JSON.stringify({ sha: 'content-sha' }), { status: 200 } ],
  getBranch: [ JSON.stringify({ commit: { sha: 'commit-sha' } }), { status: 200 }],
  setContent: BLANK_RESPONSE,
  removeContent: BLANK_RESPONSE,
  createBranch: BLANK_RESPONSE,
  deleteBranch: BLANK_RESPONSE,
  createMerge: BLANK_RESPONSE 
};

describe('gh-storage', () => {
  let github;

  beforeAll(() => {
    Math._random = Math.random;
    Math.random = () => 0.5555;
  });

  afterAll(() => {
    Math.random = Math._random;
  });

  beforeEach(() => {
    github = new GhStorage({
      repo: 'someone/something',
      data: '/data',
      uploads: '/uploads',
      branch: 'master'
    });

    github.credentials = { token: MOCK_TOKEN };
  });

  afterEach(() => {
    fetch.resetMocks();
  });

  describe('set', () => {
    it('should fetch and update files on GitHub', async () => {
      const [ mockFetch ] = fetch.mockResponses(RESPONSES.getContent, RESPONSES.setContent);
      await github.set('/foo', { some: 'data' });

      const [ get, put ] = mockFetch.mock.calls;

      expect(get).toMatchSnapshot();
      expect(put).toMatchSnapshot();
    });
  });

  describe('remove', () => {
    it('should get and remove blobs from GitHub', async () => {
      const [ mockFetch ] = fetch.mockResponses(RESPONSES.getContent, RESPONSES.removeContent);
      await github.remove('/foo');

      const [ get, remove ] = mockFetch.mock.calls;

      expect(get).toMatchSnapshot();
      expect(remove).toMatchSnapshot();
    });
  });

  describe('startTransaction', () => {
    it('should create a new branch', async () => {
      const [ mockFetch ] = fetch.mockResponses(RESPONSES.getBranch, RESPONSES.createBranch);
      await github.startTransaction();

      const [ getBranch, createBranch ] = mockFetch.mock.calls;

      expect(getBranch).toMatchSnapshot();
      expect(createBranch).toMatchSnapshot();
    });
  });

  describe('transaction', () => {
    let transaction;

    beforeEach(async () => {
      fetch.mockResponses(RESPONSES.getBranch, RESPONSES.createBranch);
      transaction = await github.startTransaction();
      fetch.resetMocks();
    });

    describe('commit', () => {
      it('should create a merge', async () => {
        const [ mockFetch ] = fetch.mockResponses(RESPONSES.createMerge, RESPONSES.deleteBranch);
        await transaction.commit();
        const [ createMerge, deleteBranch ] = mockFetch.mock.calls;

        expect(createMerge).toMatchSnapshot();
        expect(deleteBranch).toMatchSnapshot();
      });
    });
  });
});