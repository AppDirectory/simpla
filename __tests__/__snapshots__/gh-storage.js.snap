// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gh-storage remove should get and remove blobs from GitHub 1`] = `
Object {
  "auth": "token some-token",
  "body": undefined,
  "method": "GET",
  "url": "https://api.github.com/repos/someone/something/contents/data/foo.json?ref=master",
}
`;

exports[`gh-storage remove should get and remove blobs from GitHub 2`] = `
Object {
  "auth": "token some-token",
  "body": Object {
    "branch": "master",
    "message": "Removing /foo",
    "path": "data/foo.json",
    "sha": "content-sha",
  },
  "method": "DELETE",
  "url": "https://api.github.com/repos/someone/something/contents/data/foo.json",
}
`;

exports[`gh-storage set should fetch and update files on GitHub 1`] = `
Object {
  "auth": "token some-token",
  "body": undefined,
  "method": "GET",
  "url": "https://api.github.com/repos/someone/something/contents/data/foo.json?ref=master",
}
`;

exports[`gh-storage set should fetch and update files on GitHub 2`] = `
Object {
  "auth": "token some-token",
  "body": Object {
    "branch": "master",
    "content": "ewogICJzb21lIjogImRhdGEiCn0=",
    "message": "Updating /data/foo.json",
    "path": "data/foo.json",
    "sha": "content-sha",
  },
  "method": "PUT",
  "url": "https://api.github.com/repos/someone/something/contents/data/foo.json",
}
`;

exports[`gh-storage startTransaction should create a new branch 1`] = `
Object {
  "auth": "token some-token",
  "body": undefined,
  "method": "GET",
  "url": "https://api.github.com/repos/someone/something/branches/master",
}
`;

exports[`gh-storage startTransaction should create a new branch 2`] = `
Object {
  "auth": "token some-token",
  "body": Object {
    "ref": "refs/heads/simpla-editing-7di5k",
    "sha": "commit-sha",
  },
  "method": "POST",
  "url": "https://api.github.com/repos/someone/something/git/refs",
}
`;

exports[`gh-storage transaction commit should create a merge 1`] = `
Object {
  "auth": "token some-token",
  "body": Object {
    "base": "master",
    "head": "simpla-editing-7di5k",
    "message": "Merging editing branch",
  },
  "method": "POST",
  "url": "https://api.github.com/repos/someone/something/merges",
}
`;

exports[`gh-storage transaction commit should create a merge 2`] = `
Object {
  "auth": "token some-token",
  "body": undefined,
  "method": "DELETE",
  "url": "https://api.github.com/repos/someone/something/git/refs/heads/simpla-editing-7di5k",
}
`;
